pkg/webhook/webhook.go

```diff
@@ -0,0 +1,372 @@
+	hostnetworkVerb = "rio-hostnetwork"
+	privilegedVerb  = "rio-privileged"
+	hostmountVerb   = "rio-hostmount"
+	servicemeshVerb = "rio-servicemesh"
+)
+
+func SetupValidatingWebhook(rContext *types.Context) error {
+	secret, err := rContext.Core.Core().V1().Secret().Get(rContext.Namespace, constants.AuthWebhookSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	devHost := os.Getenv("LOCALHOST")
+	if devHost == "" {
+		devHost = "localhost"
+	}
+	if len(secret.Data) == 0 {
+		hostname := fmt.Sprintf("%s.%s.svc", constants.AuthWebhookServiceName, rContext.Namespace)
+		if constants.DevMode != "" {
+			hostname = devHost
+		}
+		webhookCa, err := tls.GenerateRootCAWithDefaults(hostname)
+		if err != nil {
+			return err
+		}
+		secret.Data = map[string][]byte{
+			corev1.TLSPrivateKeyKey: []byte(webhookCa.Cred.EncodePrivateKeyPEM()),
+			corev1.TLSCertKey:       []byte(webhookCa.Cred.EncodeCertificatePEM()),
+			"ca":                    []byte(webhookCa.Cred.EncodeCertificatePEM()),
+		}
+		if _, err := rContext.Core.Core().V1().Secret().Update(secret); err != nil {
+			return err
+		}
+	}
+
+	if constants.DevMode != "" {
+		tlsKeyPath = fmt.Sprintf("%s/.local/ssl/tls.key", os.Getenv("HOME"))
+		tlsCrtPath = fmt.Sprintf("%s/.local/ssl/tls.crt", os.Getenv("HOME"))
+		if err := os.MkdirAll(filepath.Dir(tlsKeyPath), 0755); err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(tlsKeyPath, secret.Data[corev1.TLSPrivateKeyKey], 0755); err != nil {
+			return err
+		}
+
+		if err := ioutil.WriteFile(tlsCrtPath, secret.Data[corev1.TLSCertKey], 0755); err != nil {
+			return err
+		}
+	}
+
+	failPolicy := v1beta1.Fail
+	validatingWebhook := &v1beta1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: constants.AuthWebhookServiceName,
+		},
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "admissionregistration.k8s.io/v1beta1",
+			Kind:       "ValidatingWebhookConfiguration",
+		},
+		Webhooks: []v1beta1.ValidatingWebhook{
+			{
+				Name: "auth-webhook.rio.io",
+				NamespaceSelector: &metav1.LabelSelector{
+					MatchExpressions: []metav1.LabelSelectorRequirement{
+						{
+							Key:      "rio.cattle.io/is-system",
+							Operator: metav1.LabelSelectorOpDoesNotExist,
+						},
+					},
+				},
+				ClientConfig: v1beta1.WebhookClientConfig{
+					Service: &v1beta1.ServiceReference{
+						Namespace: rContext.Namespace,
+						Name:      "auth-webhook",
+					},
+					CABundle: secret.Data["ca"],
+				},
+				FailurePolicy: &failPolicy,
+				Rules: []v1beta1.RuleWithOperations{
+					{
+						Operations: []v1beta1.OperationType{
+							v1beta1.Create,
+							v1beta1.Update,
+						},
+						Rule: v1beta1.Rule{
+							APIGroups:   []string{rioAPIGroup},
+							Resources:   []string{rioService},
+							APIVersions: []string{"v1"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if constants.DevMode != "" {
+		validatingWebhook.Webhooks[0].ClientConfig = v1beta1.WebhookClientConfig{
```

Comment:
If we are in dev mode, can we just empty the `rule.resources`, making everything open ?

Reply:
Follow up question, is this going to be a feature flag ?

Reply:
I think we should always enable in dev mode in case we need to test this function, but webhook actually will allow everything if you have cluster-admin for privilege. Open to discussion for feature flag...